fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => {
+        this.setState({ todos: data });
+      });
+  }
+  /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+
+  addTodo(newTodo) {
+    fetch('http://localhost:3000/api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => {
+        this.setState({ todos: this.state.todos.concat(data) });
+      });
+  }
+  /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    *
+    */
+
+  toggleCompleted(todoId) {
+    fetch(`http://localhost:3000/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({ isCompleted: !this.state.todos[todoId - 1].isCompleted })
```

Comment:
We don't use the id to calculate the index. the id is the "identity", the index is the position. these things are not related. for example, your cohort is not seated by birthday. You'll need to write some code _before_ you call `fetch()`.